apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/RowAsyncEnumerator.cs

```diff
@@ -0,0 +1,298 @@
+                    owner.Assert(
+                        owner._currentCell.Family != null,
+                        "NewCell has a qualifier, but no familyName");
+                    owner._currentCell.Family.Columns.Add(owner._currentCell.Column);
+                }
+
+                idChanged = idChanged ||
+                    owner._currentCell.Timestamp != chunk.TimestampMicros;
+                owner._currentCell.Timestamp = chunk.TimestampMicros;
+
+                idChanged = idChanged ||
+                    !chunk.Labels.SequenceEqual(owner._currentCell.Labels ?? Enumerable.Empty<string>());
+                owner._currentCell.Labels = chunk.Labels;
+
+                owner.Assert(idChanged, "NewCell must have a new identifier");
+                owner.Assert(chunk.Value != null, "NewCell must have a value");
+
+                // calculate cell size
+                owner._currentCell.ValueSizeExpected = chunk.ValueSize > 0 ? chunk.ValueSize : chunk.Value.Length;
+                owner._currentCell.ValueSizeRemaining = owner._currentCell.ValueSizeExpected - chunk.Value.Length;
+
+                // Start building cell
+                owner._currentCell.Value = chunk.Value;
+
+                if (owner._currentCell.ValueSizeRemaining != 0)
+                {
+                    return CellInProgress.Instance;
+                }
+
+                owner.AddCompletedCellToRow();
+                return NewCell.Instance;
+            }
+            
+            public override RowMergeState CommitRow(RowAsyncEnumerator owner)
+            {
+                return NewRow.Instance;
+            }
+        }
+
+        private sealed class CellInProgress : RowMergeState
+        {
+            public static readonly CellInProgress Instance = new CellInProgress();
+
+            private CellInProgress() { }
+
+            public override RowMergeState HandleChunk(RowAsyncEnumerator owner, CellChunk chunk)
+            {
+                owner.HadSplitCell = true;
+                owner.Assert(chunk.FamilyName == null, "CellInProgress can't have a family");
+                owner.Assert(chunk.Qualifier == null, "CellInProgress can't have a qualifier");
+                owner.Assert(chunk.TimestampMicros == 0, "CellInProgress can't have a timestamp");
+                owner.Assert(chunk.Labels.Count == 0, "CellInProgress can't have labels");
+                owner.Assert(chunk.Value != null, "CellInProgress must have a value");
+
+                owner._currentCell.ValueSizeRemaining -= chunk.Value.Length;
+
+                bool isLast = chunk.ValueSize == 0;
+                if (isLast)
+                {
+                    owner.Assert(
+                        owner._currentCell.ValueSizeRemaining == 0,
+                        $"CellInProgress is last, but is missing {owner._currentCell.ValueSizeRemaining} bytes");
+                }
+
+                owner._currentCell.Value = owner._currentCell.Value.Concat(chunk.Value);
```

Comment:
How often do we expect all this to happen? We could keep a sequence of values instead, and concatenate more efficiently in one go.

Reply:
Never at the moment :) The server doesn't do cell splitting currently. But yes, perhaps it would be best to avoid a O(N<sup>2</sup>) algorithm. However, I would like to avoid doing a
`ByteString => List<byte> => ByteString`
round-trip in the common case. I think I know what to do, but you should probably take another look.